test_journal.py

```diff
@@ -0,0 +1,281 @@
+    kwargs['session'] = db_session
+    # first, assert that there are no entries in the database:
+    # assert db_session.query(journal.Entry).count() == 0
+    # now, create an entry using the 'write' class method
+    entry = journal.Entry.write(**kwargs)
+    db_session.flush()
+    return entry
+
+
+def test_listing(app, entry):
+    response = app.get('/')
+    assert response.status_code == 200
+    actual = response.body
+    for field in ['title', 'content']:
+        expected = getattr(entry, field, 'absent')
+        assert expected in actual
+
+
+def test_post_to_add_view(app):
+    entry_data = {
+        'title': 'Hello there',
+        'content': 'This is a post',
+    }
+    response = app.post('/add', params=entry_data, status='3*')
+    redirected = response.follow()
+    actual = redirected.body
+    for expected in entry_data.values():
+        assert expected in actual
+
+
+def test_try_to_get(app):
+    with pytest.raises(webtest.AppError):
+        app.get('/add')
+
+
+def test_add_no_params(app):
+    response = app.post('/add', status=500)
+    assert 'IntegrityError' in response.body
+
+
+@pytest.fixture(scope='function')
+def auth_req(request):
+    manager = BCRYPTPasswordManager()
+    settings = {
+        'auth.username': 'admin',
+        'auth.password': manager.encode('secret')
+    }
+    testing.setUp(settings=settings)
+    req = testing.DummyRequest()
+
+    def cleanup():
+        testing.tearDown()
+
+    request.addfinalizer(cleanup)
+
+    return req
+
+
+def test_do_login_success(auth_req):
+    from journal import do_login
+    auth_req.params = {'username': 'admin', 'password': 'secret'}
+    assert do_login(auth_req)
+
+
+def test_do_login_bad_pass(auth_req):
+    from journal import do_login
+    auth_req.params = {'username': 'admin', 'password': 'wrong'}
+    assert not do_login(auth_req)
+
+
+def test_do_login_bad_user(auth_req):
+    from journal import do_login
+    auth_req.params = {'username': 'bad', 'password': 'secret'}
+    assert not do_login(auth_req)
+
+
+def test_do_login_missing_params(auth_req):
+    from journal import do_login
+    for params in ({'username': 'admin'}, {'password': 'secret'}):
+        auth_req.params = params
+        with pytest.raises(ValueError):
+            do_login(auth_req)
+
+
+INPUT_BTN = '<input type="submit" value="Share" name="Share"/>'
+CREATE_LINK = '<a href="http://localhost/create">CREATE</a>'   # nav bar link
+
+
+def login_helper(username, password, app):
+    """encapsulate app login for reuse in tests
+
+    Accept all status codes so that we can make assertions in tests
+    """
+    login_data = {'username': username, 'password': password}
+    return app.post('/login', params=login_data, status='*')
+
+
+def test_start_as_anonymous(app):
+    response = app.get('/', status=200)
+    actual = response.body
+    assert CREATE_LINK not in actual   # check for 'CREATE' in nav bar
```

Comment:
My tests are simple, but lame.  Bald string matching is a shitty way to test html.  the response you get back from `WebTest.TestApp` has an `html` attribute which allows you much more interesting ways to compare the html you got to the html you wanted.